packages/hydrooj/src/model/task.ts

```diff
@@ -69,11 +104,13 @@ class TaskModel {
             _id: new ObjectId(),
         };
         const res = await coll.insertOne(t);
+        bus.broadcast('task/add');
```

Comment:
I don't think `broadcast` is such a high cost operation since it would be called with `record/change` during the judge process when `next` or `end` is being called. The delay was up to 1 second before task begin processed, which is not a small in my opinion. At the same time, the polling model requires unnecessary database read when idle.

To use MongoDB as an event-triggering source, I would rather use a tail-able cursor with a capped collection for higher performance. In this case, we may choose to convert `Task` into a capped collection with huge size (e.g. 100k) or use a dedicated collection for notify purpose only.

Reply:
Compared to the compilation and judge process that usually take seconds I really don't agree that a <1sec delay is that important (and in real-world deployments you would usually get more than 1 consumer).
I would prefer wasting some database operation when idle than adding extra complexity in high server load.
Also requiring manually trigger of notify would break those users who are manually inserting documents into the task collection.
In current design, a broadcast operation will cause N database write operation while N is the number of web servers connected.

FYI, if you are worried about judge callbacks, there's also a `performance` setting for judge and if it's enabled, it will skip per-testcase callback to get better performance.

Using tailable cursor, you still have to poll for events, while collection.watch gives a native way to get the change stream. Not sure about performance difference among them.

Reply:
Okey, I'll change back to polling model for now. Maybe we can think of a hybrid implementation that unionize both polling and pull model.